src/dune_rpc_impl/server.ml

```diff
@@ -119,34 +119,97 @@ end)
-    { build_progress = Session_set.empty; diagnostics = Session_set.empty }
+  (** Remove all subscriptions of a sesssion *)
+  val remove_session : t -> Client.t Session.t -> t
 
-  let get t (which : Subscribe.t) =
-    match which with
-    | Build_progress -> t.build_progress
-    | Diagnostics -> t.diagnostics
+  val add : t -> 'a what -> 'a Subscription.t -> t
+
+  val notify : t -> 'a what -> 'a -> unit Fiber.t
+end = struct
+  module S = struct
+    let compare = Subscription.compare
+
+    let to_dyn x = Subscription.to_dyn x
+  end
 
-  let modify t (which : Subscribe.t) ~f =
-    match which with
-    | Build_progress -> { t with build_progress = f t.build_progress }
-    | Diagnostics -> { t with diagnostics = f t.diagnostics }
+  module Progress_subs_comparable = Comparable.Make (struct
+    type t = Progress.t Subscription.t
 
-  let add t which session =
-    modify t which ~f:(fun set -> Session_set.add set session)
+    include S
+  end)
 
-  let remove t which session =
-    modify t which ~f:(fun set -> Session_set.remove set session)
+  module Progress_subs = Progress_subs_comparable.Set
 
-  let notify t which ~f =
-    let set = get t which in
-    Fiber.parallel_iter_set (module Session_set) set ~f
+  module Diagnostic_subs_comparable = Comparable.Make (struct
+    type t = Diagnostic.Event.t list Subscription.t
+
+    include S
+  end)
+
+  module Diagnostic_subs = Diagnostic_subs_comparable.Set
+  module Session_map = Session_comparable.Map
+
+  module Sub_comparable = Comparable.Make (struct
+    type t = Subscription.packed
+
+    let compare x y = Subscription.compare_packed x y
+
+    let to_dyn (Subscription.E x) = Subscription.to_dyn x
+  end)
+
+  module Sub_map = Sub_comparable.Map
+
+  type t =
+    { build_progress : Progress.t Subscription.t Sub_map.t
+    ; diagnostic : Diagnostic.Event.t list Subscription.t Sub_map.t
+    }
+
+  type _ what =
+    | Diagnostic : Diagnostic.Event.t list what
+    | Progress : Progress.t what
+
+  let empty = { build_progress = Sub_map.empty; diagnostic = Sub_map.empty }
+
+  let add (type a) t (what : a what) (sub : a Subscription.t) =
+    let key = Subscription.E sub in
+    match what with
+    | Progress ->
+      { t with build_progress = Sub_map.add_exn t.build_progress key sub }
+    | Diagnostic -> { t with diagnostic = Sub_map.add_exn t.diagnostic key sub }
+
+  let remove (type a) t (what : a what) (sub : a Subscription.t) =
+    let key = Subscription.E sub in
+    match what with
+    | Progress ->
+      { t with build_progress = Sub_map.remove t.build_progress key }
+    | Diagnostic -> { t with diagnostic = Sub_map.remove t.diagnostic key }
+
+  let remove_session t session =
+    Session.subscriptions session
+    |> List.fold_left ~init:t ~f:(fun t sub ->
+           { build_progress = Sub_map.remove t.build_progress sub
+           ; diagnostic = Sub_map.remove t.diagnostic sub
+           })
+
+  let notify (type a) t (what : a what) (a : a) =
```

Comment:
Here we basically implement a pipe "forking" logic (similar to `Async.Pipe.fork`, but n-ary).
Forking pipes has an annoying property: you have to either wait for the slowest consumer, or have an unbounded buffering.
From this code it looks like it's waiting for the slowest consumer, but I'm not sure (is `Subscription.update` actually supposed to do pushback, or is it in the monad for some less interesting reason?). Can you add some comment on this?

It would be nice to have the pipe forking  logic in a separate module where this could be documented and thought about more abstractly (without having to do it separately for `Progress` and for `Diagnostics`).
I do realize there's not much code to share, so I'll understand if you say it's not worth it.

Reply:
> From this code it looks like it's waiting for the slowest consumer, but I'm not sure (is Subscription.update actually supposed to do pushback, or is it in the monad for some less interesting reason?). Can you add some comment on this?

We are going for unbounded queuing actually. We call `notify` inside a `Fiber.Pool.t` dedicated for delivering notifications.

> It would be nice to have the pipe forking logic in a separate module where this could be documented and thought about more abstractly (without having to do it separately for Progress and for Diagnostics).
I do realize there's not much code to share, so I'll understand if you say it's not worth it.

Let me have a look at `Pipe.fork` and see what can be done here. As a first reaction, I'm not eager to abstract things because I've already rewritten this code more than once.

All of this talk about pushback makes me wonder if this interface is the wrong way to go. What do you think of long polling? I have some experience with it from client side JS work, but it was always considered a hack necessitated  by the lack of proper streaming in HTTP. At least it handily avoids all the problems of pushback.

Reply:
I don't know what "long" means in "long polling".

I can see polling being a reasonable way to implement pushback, but I think we'll still have unbounded queueing, so what does it actually solve here?

> Let me have a look at `Pipe.fork`

I think pipe internals is pretty scary so I wouldn't necessarily recommend.

> I'm not eager to abstract things

Fair enough, this is fine.

Reply:
> I don't know what "long" means in "long polling".

It means that when the client makes a request for an update but the server doesn't have one to offer, it will wait until an update is available to respond. Hence the polling request might be pending for a long time.

> I can see polling being a reasonable way to implement pushback, but I think we'll still have unbounded queueing, so what does it actually solve here?

You mean queuing to serve these requests when an update arrives? This will be mostly mitigated by the fact that clients will not serve additional requests until they receive responses for the one they have pending. So we will be waiting for at 1 most 1 ivar per subscription.

Reply:
Oh, got it! Yeah, long polling sounds like a great approach, then.